test/test_models.py

```diff
@@ -0,0 +1,369 @@
+
+    def __init__(self, *args, **kwargs):
+        super(TestModel, self).__init__(*args, **kwargs)
+        self.opt = opt
+
+    # Helper to generate a vocabulary
+
+    def get_vocab(self):
+        return onmt.Dict([onmt.Constants.PAD_WORD, onmt.Constants.UNK_WORD,
+                          onmt.Constants.BOS_WORD, onmt.Constants.EOS_WORD])
+
+    def get_batch(self, sourceL=3, bsize=1):
+        # len x batch x nfeat
+        test_src = Variable(torch.ones(sourceL, bsize, 1)).long()
+        test_tgt = Variable(torch.ones(sourceL, bsize)).long()
+        test_length = Variable(torch.ones(1, bsize).fill_(sourceL))
+        return test_src, test_tgt, test_length
+
+    def test_10_embeddings_init(self):
+        try:
+            # Initialize Dictionary
+            vocab = self.get_vocab()
+            onmt.Models.Embeddings(self.opt, vocab)
+        except:
+            self.fail("Embedding initialization failed.")
+
+    def test_11_embeddings_forward(self):
+        vocab = self.get_vocab()
+        emb = onmt.Models.Embeddings(self.opt, vocab)
+
+        sourceL = 3
+        bsize = 1
+        test_src, _, __ = self.get_batch(sourceL=sourceL,
+                                         bsize=bsize)
+        res = emb(test_src)
+        compare_to = torch.zeros(sourceL, bsize, self.opt.word_vec_size)
+        self.assertEqual(res.size(), compare_to.size())
+
+    def test_12_embeddings_forward_transformer_decoder(self):
+        vocab = self.get_vocab()
+        emb = onmt.Models.Embeddings(self.opt, vocab)
+
+        sourceL = 3
+        bsize = 1
+        test_src, _, __ = self.get_batch(sourceL=sourceL,
+                                         bsize=bsize)
+        input = torch.cat([test_src, test_src], 0)
+        res = emb(input)
+        compare_to = torch.zeros(sourceL*2, bsize, self.opt.word_vec_size)
+
+        self.assertEqual(res.size(), compare_to.size())
+
+    def test_20_encoder_init(self):
+        vocab = self.get_vocab()
+        onmt.Models.Encoder(self.opt, vocab)
+
+    def encoder_forward(self, opt, sourceL=3, bsize=1):
+        vocab = self.get_vocab()
+        enc = onmt.Models.Encoder(opt, vocab)
+
+        test_src, test_tgt, test_length = self.get_batch(sourceL=sourceL,
+                                                         bsize=bsize)
+
+        hidden_t, outputs = enc(test_src, test_length)
+
+        # Initialize vectors to compare size with
+        test_hid = torch.zeros(self.opt.layers, bsize, opt.rnn_size)
+        test_out = torch.zeros(sourceL, bsize, opt.rnn_size)
+
+        # Ensure correct sizes and types
+        self.assertEqual(test_hid.size(),
+                         hidden_t[0].size(),
+                         hidden_t[1].size())
+        self.assertEqual(test_out.size(), outputs.size())
+        self.assertEqual(type(outputs), torch.autograd.Variable)
+        self.assertEqual(type(outputs.data), torch.FloatTensor)
+
+    def test_21_encoder_forward(self):
```

Comment:
Trying to order the tests because they run in alphabetical order - will no longer be necessary with declarative testing.